orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java

```diff
@@ -353,15 +356,34 @@ public void initServiceRun(List<GatewayConfig> allGateway, Set<Node> allNodes, S
             Callable<Boolean> saltPillarRunner = saltRunner.runner(hostSave, exitCriteria, exitModel);
             saltPillarRunner.call();
 
+            // Observation: The only pillar entries which seem to change when adding new nodes are disks and nodes.
+            //  PillarSave isn't very expensive - but the entire set of pillarSaves does take over 1 second.
             for (Entry<String, SaltPillarProperties> propertiesEntry : saltConfig.getServicePillarConfig().entrySet()) {
                 OrchestratorBootstrap pillarSave = new PillarSave(sc, gatewayTargetIpAddresses, propertiesEntry.getValue());
                 saltPillarRunner = saltRunner.runner(pillarSave, exitCriteria, exitModel);
                 saltPillarRunner.call();
             }
 
+            // Each grainAddRunner takes at least 0.6 seconds (combination of grains.append + grains.get). Occasionally longer than 1 second.
+            //  All of the grain additions here end up adding 5 seconds to a new node addition.
+
+            // Salt allows for multiple grains to be added in a single call. This call, howeever, will not throw an error
+            //  if a grain already exists. Instead, it will keep appending to the list of grains.
+            //  Since this operation is called with all nodes - moving over to a single call to append multiple grains
+            //  will likely break a lot of things.
+
+            // TODO Question: Back to why this is invoked with allNodes vs new nodes only. NewNodes only allows for such optimizations.
+            //  In theory, this could be a grains.get + construct the append command in node batches (i.e. existing nodes will get almost no
+            //  changes. New nodes will be sent as a single bulk operation). Not familiar enough with the flow to be confident about making
+            //  such a change. 5s is a reasonable amount of time though.
             setAdMemberRoleIfNeeded(allNodes, saltConfig, exitModel, sc, allNodeHostname);
             setIpaMemberRoleIfNeeded(allNodes, saltConfig, exitModel, sc, allNodeHostname);
 
+            // TODO Question: Can some of these grains be set ahead of time during image building, or salt-bootstrap to
```

Comment:
We use the same image for all instances and all types of clusters. DL & DH, thus we can't distinguish in image build time.